oprf/group/group.go

```diff
@@ -0,0 +1,206 @@
+
+	return r
+}
+
+// Neg performs the negation operation on the calling Element object
+func (p *Element) Neg() *Element {
+	xInv := new(big.Int).ModInverse(p.x, p.c.curve.Params().N)
+
+	r := NewElement(p.c)
+	r.x.Set(xInv)
+	r.y.Set(p.y)
+
+	return r
+}
+
+// Serialize the Element into a byte slice.
+func (p *Element) Serialize() []byte {
+	x, y := p.x.Bytes(), p.y.Bytes()
+
+	// append zeroes to the front if the bytes are not filled up
+	x = append(make([]byte, p.c.ByteLength()-len(x)), x...)
+	y = append(make([]byte, p.c.ByteLength()-len(y)), y...)
+
+	var b []byte
+	var tag int
+	if !p.com {
+		b = append(x, y...)
+		tag = 4
+	} else {
+		b = x
+		sign := Sgn0(p.y)
+		// select correct tag
+		e := int(Equals(sign, One).Int64())
+		tag = subtle.ConstantTimeSelect(e, 2, 3)
+	}
+
+	return append([]byte{byte(tag)}, b...)
+}
+
+// checkBytes checks that the number of bytes corresponds to the correct
+// curve type and serialization tag that is present
+func isCompressed(in []byte, expLen int) (bool, error) {
+	tag := in[0]
+	com := false
+
+	switch tag {
+	case 2, 3:
+		if expLen < len(in)-1 {
+			return false, errors.New("error deserializing group element")
+		}
+		com = true
+	case 4:
+		if expLen*2 < len(in)-1 {
+			return false, errors.New("error deserializing group element")
+		}
+	default:
+		return false, errors.New("error deserializing group element")
+	}
+
+	return com, nil
+}
+
+// Deserialize an octet-string into a valid Element object.
+func (p *Element) Deserialize(in []byte) error {
+	byteLen := p.c.ByteLength()
+
+	com, err := isCompressed(in, byteLen)
+	if err != nil {
+		return err
+	}
+
+	if !com {
+		p.x = new(big.Int).SetBytes(in[1 : byteLen+1])
+		p.y = new(big.Int).SetBytes(in[byteLen+1:])
+
+		if !p.IsValid() {
+			return errors.New("invalid deserialization")
+		}
+	} else {
+		order := p.c.curve.Params().P
+		var y2 *big.Int
+		x := new(big.Int).SetBytes(in[1:])
+
+		x2 := new(big.Int).Exp(x, Two, order)
+		x2a := new(big.Int).Add(x2, big.NewInt(-3))
+		x3 := new(big.Int).Mul(x2a, x)
+		x3ab := new(big.Int).Add(x3, p.c.curve.Params().B)
+		y2 = new(big.Int).Mod(x3ab, order)
+
+		a1 := new(big.Int).Add(p.c.curve.Params().P, One)
+		a2 := new(big.Int).ModInverse(big.NewInt(4), p.c.curve.Params().P)
+		m := new(big.Int).Mul(a1, a2)
+		sqrtExp := new(big.Int).Mod(m, p.c.curve.Params().P)
+
+		// construct y coordinate with correct sign
+		y := new(big.Int).Exp(y2, sqrtExp, order)
+		parity := Equals(big.NewInt(int64(in[0])), Two)
+		yParity := Equals(Sgn0(y), One)
+		y = cMov(new(big.Int).Mul(y, MinusOne), y, Equals(parity, yParity))
+
+		p.x = new(big.Int).Mod(x, order)
+		p.y = new(big.Int).Mod(y, order)
+
+		if !p.IsValid() {
+			return errors.New("invalid deserialization")
+		}
+
+		p.com = true
+	}
+
+	return nil
+}
+
+// Equal returns a bool indicating whether two Elements are equal.
+func (p *Element) Equal(q *Element) bool {
+	return (p.x.Cmp(q.x) == 0) && (p.y.Cmp(q.y) == 0)
```

Comment:
Mmm.. the `com` field is irrelevant to compare as it can be set to true or to false, and yet the internal representation of the Element will be equal, as compressed or uncompressed it will be the same Element. The `com` field is only for serialization or deserialization purposes. The Ciphersuite is just for compatibility for OPRF, but we can check it (although in the future for the Group API this will not be included).

Reply:
With the new change, I don't think it is needed to compare the interface `elliptic.curve`.